javascript/MaterialXView/source/viewer.js

```diff
@@ -623,15 +667,110 @@ export class Material
         }
         else
         {
-            // Search for any surface shader. It
+            // Search for any surface shaders. The first found
             // is assumed to be assigned to the entire scene
             // The identifier used is "*" to mean the entire scene. 
-            let elem = mx.findRenderableElement(doc);
-            if (elem)
+            const materialNodes = doc.getMaterialNodes();
+            let shaderList = [];
+            let foundRenderable = false;            
+            for (let i=0; i<materialNodes.length; ++i)
+            {
+                let materialNode = materialNodes[i];
+                if (materialNode)
+                {
+                    console.log('Scan material: ', materialNode.getNamePath());
+                    let shaderNodes = mx.getShaderNodes(materialNode)
+                    if (shaderNodes.length > 0)
+                    {
+                        let shaderNodePath = shaderNodes[0].getNamePath()
+                        if (!shaderList.includes(shaderNodePath))
+                        {
+                            let assignment = NO_GEOMETRY_SPECIFIER;
+                            if (foundRenderable == false)
+                            {
+                                assignment = ALL_GEOMETRY_SPECIFIER;
+                                foundRenderable = true;
+                            }                        
+                            console.log('-- add shader: ', shaderNodePath);
+                            shaderList.push(shaderNodePath);
+                            this._materials.push(new MaterialAssign(shaderNodes[0], assignment));
+                        }
+                    }
+                }            
+            }
+            const nodeGraphs = doc.getNodeGraphs();
+            for (let i=0; i<nodeGraphs.length; ++i)
+            {
+                let nodeGraph = nodeGraphs[i];
+                if (nodeGraph)
+                {
+                    if (nodeGraph.hasAttribute('nodedef') || nodeGraph.hasSourceUri())
+                    {
+                        continue;
+                    } 
+                    // Skip any nodegraph that is connected to something downstream
+                    if (nodeGraph.getDownstreamPorts().length > 0)
```

Comment:
This needs the missing API exposure. Prevents re-including nodegraphs which are already connected to downstream shaders.